### PR TITLE
gui: Add Windows taskbar progress

### LINKED ISSUE #14134: windows: Show sync progress in task bar
See initial discussion in #14090. This should probably be implemented using `QWinTaskbarProgress`, and could include some build system changes to include `winextras`.

### PR DESCRIPTION
**Pull request by ken2812221**
ref: #14090 
Close #14134 

This PR adds an progress bar at taskbar on Windows. This progress bar will hide when full synced.

### CODE REVIEW

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -886,6 +893,19 @@ void BitcoinGUI::setNumBlocks(int count, const QDateTime& blockDate, double nVer
...
     progressBarLabel->setToolTip(tooltip);
     progressBar->setToolTip(tooltip);
+
+#ifdef Q_OS_WIN
+    if (m_taskbar_button->window() == nullptr) {
+        m_taskbar_button->setWindow(windowHandle());
+    }
+    QWinTaskbarProgress* taskbar_progress = m_taskbar_button->progress();
+    if (nVerificationProgress < 1.0) {
+        taskbar_progress->setValue(100.0 * nVerificationProgress + 0.5);
```
Comment (luke-jr) [inline src/qt/bitcoingui.cpp]: Why `+ 0.5`? (add a comment)

**Review by luke-jr — COMMENTED** [review]
Should this also show splashscreen progress?

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (ken2812221) [reply src/qt/bitcoingui.cpp]: This is as same as here , I guess this is to round the value
https://github.com/bitcoin/bitcoin/blob/68f3c7eb080e461cfeac37f8db7034fe507241d0/src/qt/bitcoingui.cpp#L856

Comment (luke-jr) [reply src/qt/bitcoingui.cpp]: Hm, okay.

**Comment by fanquake** [comment]
Concept ACK. Thanks for following up on this so quickly, and using `QWinTaskbarProgress `.

Given that this is an entirely new implementation, and includes the build system changes, I don't think you need to cherry-pick/include the other committer.

I'll test on a Windows VM shortly.

**Comment by jonasschnelli** [comment]
Concept ACK (haven't looked at the code though)

File: src/qt/bitcoingui.h [hunk 2]
```diff
@@ -19,6 +19,9 @@
 #include <QMenu>
 #include <QPoint>
 #include <QSystemTrayIcon>
+#ifdef Q_OS_WIN
+#include <QWinTaskbarButton>
```
Comment (promag) [inline src/qt/bitcoingui.h]: Use follow declaration instead?

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (promag) [reply src/qt/bitcoingui.cpp]: Use `qRound`?

**Review by promag — COMMENTED** [review]
Much better, thanks!

File: src/qt/bitcoingui.h (hunk 2 above)
Comment (promag) [reply src/qt/bitcoingui.h]: Ops, forward declaration.

Comment (ken2812221) [reply src/qt/bitcoingui.h]: Why?

Comment (donaloconnor) [reply src/qt/bitcoingui.h]: Because you don't need to pull in the entire header file (+ what ever it includes) when using just a pointer. It's preferred in case like this to reduce compile times etc. Include it in CPP instead.

Comment (promag) [reply src/qt/bitcoingui.h]: ☝️also, look in other headers, it's already a recurring practice.

Comment (ken2812221) [reply src/qt/bitcoingui.h]: Oh, I saw that QT_NAMESPACE. I'll do that after gitian build finished too.

**Comment by ken2812221** [comment]
I have no idea what's going on in gitian build. I am able to build this.

**Comment by ken2812221** [comment]
`a99f9caf92346d9c0bf823b2e3c2bcc44f7d9c6a6205050b985e63f5e3ed0ee1` [bitcoin-0.17.99-win32-debug.zip](https://transfer.sh/KP9KW/bitcoin-0.17.99-win32-debug.zip)
`044347527dd20eb4081cdbc9acd8adc05a97925854169affcf191b915abeef8d` [bitcoin-0.17.99-win32-setup-unsigned.exe](https://transfer.sh/gbVAz/bitcoin-0.17.99-win32-setup-unsigned.exe)
`61b745860cd74c0fb93c8cc1fbfae897589b88084f3d298688ff4ec16e4782f1` [bitcoin-0.17.99-win32.zip](https://transfer.sh/U03gU/bitcoin-0.17.99-win32.zip)
`fbbc87b50c16cd027aeb9a1ccb5a16320c414f92252c4312b605223cb4673e7a` [bitcoin-0.17.99-win64-debug.zip](https://transfer.sh/16l6V7/bitcoin-0.17.99-win64-debug.zip)
`9b2b7536863379e4977d8dcf80486ee2343e60a7fda5a20c39aaff03ba66b5c2` [bitcoin-0.17.99-win64-setup-unsigned.exe](https://transfer.sh/y9h9T/bitcoin-0.17.99-win64-setup-unsigned.exe)
`50ab47bbb18eae55fc994b3d5b274208cbeae4e379b75092e34501b4aa35d485` [bitcoin-0.17.99-win64.zip](https://transfer.sh/rIdrG/bitcoin-0.17.99-win64.zip)

This is my gitian build result for 954d19aff18dc128e3208275b65facd7b55cd852 using `./gitian-build.py -bnDpj4 -m6000 -ow 0 14137`

**Comment by MarcoFalke** [comment]
I may have to purge the cache? Did a `rm -r  gitian-builder/cache/`

**Comment by MarcoFalke** [comment]
@jonasschnelli Could you try to see if your gitian build script also fails here?

**Comment by ken2812221** [comment]
I may know what's going on. makefile would check only the first source file https://github.com/bitcoin/bitcoin/blob/adf27b531a7fcd7066ab6649e8073bd1895a823a/depends/Makefile#L153-L158

**Comment by ken2812221** [comment]
@MarcoFalke This should work again

**Comment by donaloconnor** [comment]
tACK https://github.com/bitcoin/bitcoin/pull/14137/commits/c3c00220a86e5bbbfe7a494d323388edbf9c7803 on Win10

image

**Comment by jonasschnelli** [comment]
Tested on Win10 via gitian build https://bitcoin.jonasschnelli.ch/build/797 but can't see the status bar:

bildschirmfoto 2018-09-25 um 20 42 55

**Comment by ken2812221** [comment]
The status bar would hide on 100%. Should I make it always be visible?

**Comment by jonasschnelli** [comment]
@ken2812221: Okay. I see. Any good test plan how to test this on regtest?

**Comment by ken2812221** [comment]
> Okay. I see. Any good test plan how to test this on regtest?

Hmm. I don't think there is a good way to make it not to be 100% on regtest. I would have to use testnet and `-reindex` to test it.

**Comment by MarcoFalke** [comment]
I guess you'd have to set nTime of the blocks to last year or so?

**Comment by ken2812221** [comment]
Update: Now the taskbar progress show/hide along with the modaloverlay. It can be easily tested on regtest.

File: depends/Makefile [hunk 3]
```diff
@@ -159,7 +159,8 @@ endef
 
 define check_or_remove_sources
   mkdir -p $($(package)_source_dir); cd $($(package)_source_dir); \
-  test -f $($(package)_fetched) && ( $(build_SHA256SUM) -c $($(package)_fetched) >/dev/null 2>/dev/null || \
+  test -f $($(package)_fetched) && ( test `cat $($(package)_fetched) | wc -l` -eq $(words $($(package)_all_sources)) && \
+    $(build_SHA256SUM) -c $($(package)_fetched) >/dev/null 2>/dev/null || \
```
Comment (fanquake) [inline depends/Makefile]: Why is this change required?

Comment (ken2812221) [reply depends/Makefile]: See above. The current build system only detect the first file is changed or not, so winextra would not be downloaded in you have the source file cache, but it would be re-built(which cause an error). This make the Makefile detect the file amount to be exactly match.

File: depends/packages/qt.mk [hunk 4]
```diff
@@ -103,21 +107,25 @@ endef
...
   $(build_SHA256SUM) -c $($(package)_extract_dir)/.$($(package)_file_name).hash && \
   mkdir qtbase && \
   tar --strip-components=1 -xf $($(package)_source) -C qtbase && \
   mkdir qttranslations && \
   tar --strip-components=1 -xf $($(package)_source_dir)/$($(package)_qttranslations_file_name) -C qttranslations && \
   mkdir qttools && \
-  tar --strip-components=1 -xf $($(package)_source_dir)/$($(package)_qttools_file_name) -C qttools
+  tar --strip-components=1 -xf $($(package)_source_dir)/$($(package)_qttools_file_name) -C qttools && \
+  mkdir qtwinextras && \
+  tar --strip-components=1 -xf $($(package)_source_dir)/$($(package)_qtwinextras_file_name) -C qtwinextras
```
Comment (MarcoFalke) [inline depends/packages/qt.mk]: Needs rebase to add the `--no-same-owner`

File: depends/packages/qt.mk [hunk 5]
```diff
@@ -16,8 +16,12 @@ $(package)_qttranslations_sha256_hash=9822084f8e2d2939ba39f4af4c0c2320e45d599676
 $(package)_qttools_file_name=qttools-$($(package)_suffix)
 $(package)_qttools_sha256_hash=50e75417ec0c74bb8b1989d1d8e981ee83690dce7dfc0c2169f7c00f397e5117
 
+$(package)_qtwinextras_file_name=qtwinextras-$($(package)_suffix)
+$(package)_qtwinextras_sha256_hash=794090446d33eecf9eb3f3a4f6f13b3e7ab5307936f8d21e4a76e7b6ba79125d
```
Comment (luke-jr) [inline depends/packages/qt.mk]: This hash needs to be updated for the current version.

**Comment by fanquake** [comment]
Closing as up for grabs. @sipsorcery, @NicolasDorier either of you might be interested in picking this up if you think it will be an improvement to Qt on Windows.

**Comment by NicolasDorier** [comment]
I will probably pick it up once we get https://github.com/bitcoin/bitcoin/pull/15529
